Add elevation scale

jameskoster opened the pull request:
<img width="1312" alt="elevation" src="https://github.com/user-attachments/assets/efc72162-7567-4497-8bfb-d98ffaac49bc">

This PR adds variables for the elevation scale as outlined in https://github.com/WordPress/gutenberg/issues/63757. It follows the method applied in https://github.com/WordPress/gutenberg/pull/64007 to add the radius scale. 

* Adds new variables and guidelines
* Soft-deprecates old variables (`$shadow-popover` and `$shadow-modal`) by mapping them to the new ones

### Elevation scale

There are four values in the scale:

* `$elevation-x-small`: For sections and containers that group related content and controls, which may overlap other content. Example: Preview Frame.
* `$elevation-small`: For components that provide contextual feedback without being intrusive. Generally non-interruptive. Example: Tooltips, Snackbar.
* `$elevation-medium`: For components that offer additional actions. Example: Menus, Command Palette
* `$elevation-large`: For components that confirm decisions or handle necessary interruptions. Example: Modals.

In terms of values, I used the existing `$shadow-popover` for `$elevation-x-small`, and `$shadow-modal` for `$elevation-large`. `$elevation-small` and `$elevation-medium` are new values that I created by hand to slot between the existing values. It's possible these will change when it comes to applying the new scale across components. 

### Application
| `$elevation-x-small` | `$elevation-small` | `$elevation-medium` | `elevation-large` |
| --- | --- | --- | --- |
| Content / Preview frame<br>ZStack<br>Potential input elements such as; Range handle, Active item in ToggleGroupControl, Toggle handle<br>Resize handle<br>Block toolbar | Snackbar<br>Tooltip | DropdownMenu<br>Popover<br>Command Palette | Modal<br>ConfirmDialog<br>Guide |

jameskoster committed:
Add elevation scale

jameskoster commented:
> Let's make sure we have clear instructions on how to translate the new scale to existing components like [Elevation](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-elevation--docs) and the components that consume it (like [Card](https://wordpress.github.io/gutenberg/?path=/docs/components-card--docs)) and [ZStack](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-zstack--docs)

Would it make sense to simply map each step in the scale to the `value` prop in `Elevation`? Anything above 4 would just use the value associated with the fourth step.

> Let's review the Popover component and its usage

Are you asking which value would be applied to Popovers? Generally they contain un-interruptive configuration UI's, so `$elevation-30` probably makes most sense. It would be good to add well-defined guidelines around when to use popovers.

> How does the radii and elevation work relate to the @wordpress/theme effort?

That's a good question. In terms of elevation the shadow styles will need to be different for light / dark modes. Darker colored themes will require darker shadow styles in order for them to be visible. 

I don't _think_ radius should be affected, but another consideration is spacing; the theme component may need to support a 'density' option that affects spacing, type styles, components, etc. That part isn't really clear yet.

ciampo commented:
> Would it make sense to simply map each step in the scale to the `value` prop in `Elevation`? Anything above 4 would just use the value associated with the fourth step.

That could work, although it could create unexpected changes in consumers. If we went down that way, we'd need to make sure to communicate it to consumers of those components via a dev note.

> Are you asking which value would be applied to Popovers? Generally they contain un-interruptive configuration UI's, so $elevation-30 probably makes most sense. It would be good to add well-defined guidelines around when to use popovers.

Absolutely. Changing the default popover box shadow is trivial, but what I really meant was that this is a great opportunity to review all popovers across Gutenberg and standardize their box shadow for better consistency and clarity.

tyxla commented:
> Would it make sense to simply map each step in the scale to the value prop in Elevation? Anything above 4 would just use the value associated with the fourth step.

I think this would create discrepancies with the existing styles of `Elevation`. Perhaps we could provide a different syntax for the 4 scale values - for example, similar to the radius scale variable names in #64007?

jasmussen commented:
Just capturing some small feedback/questions from [here](https://github.com/WordPress/gutenberg/issues/63757#issuecomment-2259813272), relevant to the PR.

richtabor commented:
Is this more of a system for elevation, or proposing what those elevations should be, or both?

jameskoster commented:
> Perhaps we could provide a different syntax for the 4 scale values - for example, similar to the radius scale variable names in https://github.com/WordPress/gutenberg/pull/64007?

That could work, although the proposed naming convention might lead to confusion? Would it be strange if `<Elevation value={ 1 } />` produced a stronger shadow than `<Elevation value={ 10 } />`? Or are you suggesting something different?

> Is this more of a system for elevation, or proposing what those elevations should be, or both?

The former, to avoid changing anything in production for now. We can define and apply values separately once we're happy with the scale.

@jasmussen If we want fewer steps in the scale I'd be tempted to combine `10` and `20`. We should probably work on some mockups (with rough shadow values) to help decide this.

jasmussen commented:
> If we want fewer steps in the scale I'd be tempted to combine 10 and 20. We should probably work on some mockups (with rough shadow values) to help decide this.

Seems fine with me. 

I'm always just in the category of _start with as little as possible_ and then expand only when the need very clearly presents itself.

hanneslsm commented:
I don't think it's currently needed because it doesn't fit the common logic of elevation (Level 0 and up) and the current design style (would be more skeuomorphic) but want to bring it up anyways, just to be sure:
negative elevation / inset shadow / depth. That'd influence the naming.
Examples could be iFrame or input fields or even buttons.

tyxla commented:
> > Perhaps we could provide a different syntax for the 4 scale values - for example, similar to the radius scale variable names in https://github.com/WordPress/gutenberg/pull/64007?
> 
> That could work, although the proposed naming convention might lead to confusion? Would it be strange if <Elevation value={ 1 } /> produced a stronger shadow than <Elevation value={ 10 } />? Or are you suggesting something different?

My thoughts were around using string `small`, `medum`, `large`, etc. for those 4 particular sizes, and those would convert to their corresponding `Elevation` values. In addition, `Elevation` would continue supporting number values as before.

jameskoster commented:
@hanneslsm shadow styles are already being applied albeit sporadically without well defined intent. 

A big part of this effort is about defining how/when to use elevation as we embark on the broader admin design effort (https://github.com/WordPress/gutenberg/issues/53322). Please refer to the four suggested steps in the scale which details this. 

In terms of mockups it's a little bit tricky because there are as yet undecided details to consider. For instance details like radius, color schemes, type, light/dark modes etc. With that said consider the following rough in terms of those details (including the shadows); they're primarily to demonstrate the 4 steps in the scale in some existing flows and experiences.

### `$elevation-10` applied to content frame.

<img width="1341" alt="Elevation 1" src="https://github.com/user-attachments/assets/1e51eeb4-1211-49bb-9dfe-74abee877722">

### `$elevation-20` applied to tooltip, snackbar, bulk-actions toolbar.
<img width="1341" alt="Elevation 2" src="https://github.com/user-attachments/assets/9050de96-16cb-41f4-b2f0-5b6a875ce058">

### `$elevation-30` applied to menu.
<img width="1341" alt="Elevation 3" src="https://github.com/user-attachments/assets/2698520f-bc41-4cab-962f-bee0e27afd9c">

### `$elevation-40` applied to modal.
<img width="1341" alt="Elevation 4" src="https://github.com/user-attachments/assets/c77dd279-3e99-4280-beee-a250fe9fa43f">

---

> My thoughts were around using string small, medum, large, etc. 

@tyxla it can work, but leaves less room for manoeuvrability if we later need to add steps in the scale.

hanneslsm commented:
> @hanneslsm shadow styles are already being applied albeit sporadically without well defined intent.

> A big part of this effort is about defining how/when to use elevation as we embark on the broader admin design effort (https://github.com/WordPress/gutenberg/issues/53322). Please refer to the four suggested steps in the scale which details this.

Thanks for clarifying! Looks good!
In terms of naming I'm also in favor of the numeric scale by the way.

jameskoster committed:
Update names

jameskoster committed:
Merge branch 'trunk' into add/elevation-scale

jameskoster commented:
I don't think there's a huge amount in it, but I've switched the names to match the radius variables, IE to use `-small`, `-medium` suffixes. Reason being:

* It's more human readable.
* Maps more cleanly to the existing `Elevation` component.
* It seems unlikely we'll need to add more steps in the future, but if necessary values like `$elevation-x-large` can be added, and components remapped as required.
* Consistency with radius

I also added some examples of where the different values would be applied across the existing set of components. 

`/packages/components/src/utils/config-values.js` is also updated.

jameskoster committed:
Update config values

DaniGuardiola reviewed (approved):
This looks correct to me from a technical standpoint. The only concern I have is that the naming might be confusing since it doesn't contain "shadow", yet it's supposed to be an option for shadow property values. However, it's not a big deal so not going to block on that.

Appreciate the comments with added context on usage of each option, it's a nice plus.

DaniGuardiola reviewed (changes requested):
Ah, missed that this is currently broken because `$shadow-popover` is assigned `$elevation-10` (similar story with "modal"), guess it's just out of sync after recent changes in naming. After this is corrected, my previous (approval) review stands. Example failure in CI: https://github.com/WordPress/gutenberg/actions/runs/10249250569/job/28352231809?pr=64108

<img width="749" alt="image" src="https://github.com/user-attachments/assets/7bda2ebb-9c96-49b5-a0bc-1e13a141d535">

jameskoster committed:
variable mapping

jameskoster commented:
Good spot, I fixe that.

jameskoster commented:
@DaniGuardiola @ciampo do you think this needs a components changelog entry?

DaniGuardiola commented:
@jameskoster good question. @ciampo and I discussed and we both think it's not necessary.

DaniGuardiola reviewed (approved):
Issue was addressed 👍

ciampo commented on packages/components/src/utils/config-values.js:
```diff
@@ -74,6 +74,10 @@ export default Object.assign( {}, CONTROL_PROPS, TOGGLE_GROUP_CONTROL_PROPS, {
 	cardPaddingMedium: `${ space( 4 ) } ${ space( 6 ) }`,
 	cardPaddingLarge: `${ space( 6 ) } ${ space( 8 ) }`,
 	popoverShadow: `0 0.7px 1px rgba(0, 0, 0, 0.1), 0 1.2px 1.7px -0.2px rgba(0, 0, 0, 0.1), 0 2.3px 3.3px -0.5px rgba(0, 0, 0, 0.1)`,
```
Since I don't think that `popoverShadow`  is exported from the package, I think it should be OK to remove it and replace its usages with `elevationXSmall`.

This can be definitely done in one of the follow-up PRs

jameskoster commented:
Cool. I'll go ahead and merge this. We can continue to refine the scale and values as needed when we begin applying across components.
